src/grpc/collections.rs

```diff
@@ -236,35 +286,55 @@ impl CollectionService for CollectionServiceImpl {
             "Unauthorized"
         );
 
-        let collection = match tonic_internal!(
-            self.database_handler.update_keyvals(request).await,
+        let collection = tonic_internal!(
+            self.database_handler.update_dataclass(request).await,
             "Internal database error."
-        ) {
-            generic_resource::Resource::Collection(c) => Some(c),
-            _ => return Err(tonic::Status::unknown("This should not happen.")),
+        );
+        self.cache
+            .update_object(&collection.object.id, collection.clone());
+        let collection: generic_resource::Resource =
+            tonic_internal!(collection.try_into(), "Collection conversion error");
+        let response = UpdateCollectionDataClassResponse {
+            collection: Some(collection.into_inner()?),
         };
-
-        Ok(tonic::Response::new(UpdateCollectionKeyValuesResponse {
-            collection,
-        }))
-    }
-
-    async fn get_collections(
-        &self,
-        _request: Request<GetCollectionsRequest>,
-    ) -> Result<Response<GetCollectionsResponse>> {
-        todo!()
-    }
-    async fn delete_collection(
-        &self,
-        _request: Request<DeleteCollectionRequest>,
-    ) -> Result<Response<DeleteCollectionResponse>> {
-        todo!()
+        return_with_log!(response);
     }
     async fn snapshot_collection(
         &self,
-        _request: Request<SnapshotCollectionRequest>,
+        request: Request<SnapshotCollectionRequest>,
     ) -> Result<Response<SnapshotCollectionResponse>> {
-        todo!()
+        log_received!(&request);
+
+        let token = tonic_auth!(
+            get_token_from_md(request.metadata()),
+            "Token authentication error."
+        );
+
+        let request = SnapshotRequest::Collection(request.into_inner());
+        let collection_id = tonic_invalid!(request.get_id(), "Invalid dataset id.");
+        let ctx = Context::res_ctx(collection_id, DbPermissionLevel::ADMIN, true);
+
+        tonic_auth!(
+            self.authorizer.check_permissions(&token, vec![ctx]).await,
+            "Unauthorized"
+        );
+
+        // this only contains one entry with a dataset
+        let resources = tonic_internal!(
+            self.database_handler.snapshot(request).await,
+            "Internal database error."
+        );
+        for resource in &resources {
+            self.cache
+                .update_object(&resource.object.id, resource.clone());
+        }
+        let collection: generic_resource::Resource =
+        // First entry is always the snapshot collection
+            tonic_internal!(resources[0].clone().try_into(), "Dataset conversion error");
```

Comment:
Copy & paste only partially adapted.

```suggestion
        let collection_id = tonic_invalid!(request.get_id(), "Invalid collection id.");
        let ctx = Context::res_ctx(collection_id, DbPermissionLevel::ADMIN, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        // This contains all resources the snapshot includes below the collection
        let resources = tonic_internal!(
            self.database_handler.snapshot(request).await,
            "Internal database error."
        );
        for resource in &resources {
            self.cache
                .update_object(&resource.object.id, resource.clone());
        }
        let collection: generic_resource::Resource =
        // First entry is always the snapshot collection
            tonic_internal!(resources[0].clone().try_into(), "Collection conversion error");
```